File: src/main/java/ru/xw1w1/lookwhatyoudone/data/DamageData.java
package ru.xw1w1.lookwhatyoudone.data;

import java.text.DecimalFormat;
import org.bukkit.Location;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

public record DamageData(
        @NotNull DamageType type, @NotNull ItemStack weapon, @NotNull Entity damaged,
        @NotNull Player damageSource, double damage, boolean critical
) {

    public @NotNull String formattedDamage() {
        return (new DecimalFormat("0.00")).format(this.damage);
    }

    public @NotNull String formattedHearts() {
        // Count of small (1/2) hearts           full heart
        int smallHearts = Math.floorMod((long) this.damage, 2);
        int bigHearts = (int) (damage - smallHearts)/2;
        return "❤".repeat(Math.max(0, bigHearts))
                .concat("♥".repeat(Math.max(0, smallHearts)));
    }

    public @NotNull Location indicatorLocation() {
        return damaged.getLocation().add(0, 1, 0);
    }

    @Override
    public String toString() {
        return this.formattedDamage();
    }

}

File: src/main/kotlin/org/ttlzmc/showdamage/datatypes/DamageData.kt
package org.ttlzmc.showdamage.datatypes

import org.bukkit.Location
import org.bukkit.damage.DamageSource
import org.bukkit.entity.Entity

data class DamageData (
    private val source: DamageSource,
    private val reciever: Entity,
    private val damageDealt: Double,
    ){
    private val valid: Boolean = BukkitDamageMask.isValidDamageType(source.damageType)
    private val location = pushLocation(reciever.location, y = 5.0)

    fun isValid(): Boolean {
        return this.valid;
    }

    private fun pushLocation(location: Location, x: Double = 0.0, y: Double = 0.0, z: Double = 0.0) {
        location.add(x, y, z)
    }


}
